src/computability/partrec_code.lean

```diff
@@ -506,6 +512,19 @@ theorem curry_prim : primrec₂ curry :=
 comp_prim.comp primrec.fst $
 pair_prim.comp (const_prim.comp primrec.snd) (primrec.const code.id)
 
+theorem one_one_curry {c₁ c₂ n₁ n₂} (h : curry c₁ n₁ = curry c₂ n₂) : c₁ = c₂ ∧ n₁ = n₂ :=
+⟨by injection h, by { injection h, 
+                      injection h with h₁ h₂, 
+                      injection h₂ with h₃ h₄, 
+                      exact injective_const h₃ }⟩
+
+theorem injective_curry : function.injective curry :=
+λ a b h, (@one_one_curry _ _ 0 0 $ by rw h).1
+
+theorem s_m_n (c n x) : ∃ f : code → ℕ → code, 
+  function.injective f ∧ computable₂ f ∧ eval (f c n) x = eval c (mkpair n x) := 
```

Comment:
Also, n and x are quantified in the wrong place here. They should be in the third conjunct, since we want to say that f has a certain behavior on any input values n,x.

Reply:
Ah, that's a shame. I copy-pasted `eval_curry` without thinking about it.

Reply:
Recursion theorists usually want it to be injective for better reducibility purposes, but as you said it's unnatural to have it relative to one of its arguments. I have removed the condition, but left the injectivity as a separate theorem.